From ba59922386144e84209b8d016c933f634590d145 Mon Sep 17 00:00:00 2001
From: Jamie Mansfield <dev@jamierocks.uk>
Date: Sat, 5 Nov 2016 21:04:09 +0000
Subject: [PATCH] Update dependencies


diff --git a/build.gradle b/build.gradle
index b8aae240..504ff52d 100644
--- a/build.gradle
+++ b/build.gradle
@@ -39,8 +39,8 @@ dependencies {
     compile 'org.jdom:jdom2:2.0.5'
     compile 'com.mchange:c3p0:0.9.5'
     compile 'org.apache.logging.log4j:log4j-api:2.0-beta9'
-    compile 'com.google.guava:guava:15.0'
-    compile 'com.mojang:authlib:1.5.17'
+    compile 'com.google.guava:guava:17.0'
+    compile 'com.mojang:authlib:1.5.21'
     compile 'com.googlecode.json-simple:json-simple:1.1.1'
     compile 'org.mcstats.standalone:metrics:R7'
     runtime 'org.xerial:sqlite-jdbc:3.8.7'
-- 
2.11.1

