From b9ebbb41575d1361a72afc503dab3ae0bacfa15f Mon Sep 17 00:00:00 2001
From: Jamie Mansfield <dev@jamierocks.uk>
Date: Tue, 3 Jan 2017 00:50:41 +0000
Subject: [PATCH] Make world auto-save interval configurable


diff --git a/src/main/java/net/canarymod/config/ServerConfiguration.java b/src/main/java/net/canarymod/config/ServerConfiguration.java
index 508722ad..65c1299c 100644
--- a/src/main/java/net/canarymod/config/ServerConfiguration.java
+++ b/src/main/java/net/canarymod/config/ServerConfiguration.java
@@ -182,6 +182,12 @@ public class ServerConfiguration implements ConfigurationContainer {
         cfg.setComments("bungeecord", "If you want to enable Bungeecord support. REQUIRES THAT ONLINE MODE IS DISABLED (false)");
         cfg.getBoolean("verify-user-session-ip", true);
         cfg.setComments("verify-user-session-ip", "Enable IP verification for connecting players");
+
+        // Neptune - start
+        cfg.setInt("world-auto-save-interval", 900);
+        cfg.setComments("world-auto-save-interval", "The number of ticks between each auto-save.");
+        // Neptune - end
+
         cfg.save();
     }
 
@@ -699,4 +705,10 @@ public class ServerConfiguration implements ConfigurationContainer {
     public boolean verifyUserSessionIP() {
         return cfg.getBoolean("verify-user-session-ip", true);
     }
+
+    // Neptune - start
+    public int getWorldAutoSaveInterval() {
+        return this.cfg.getInt("world-auto-save-interval", 900);
+    }
+    // Neptune - end
 }
-- 
2.14.1

