From faa2494a5117e680bdab632448ed90da8fcb242e Mon Sep 17 00:00:00 2001
From: Jamie Mansfield <dev@jamierocks.uk>
Date: Sat, 29 Oct 2016 20:51:19 +0100
Subject: [PATCH] Moan about Sponge plugins or Forge mods


diff --git a/src/main/java/net/canarymod/plugin/PluginDescriptor.java b/src/main/java/net/canarymod/plugin/PluginDescriptor.java
index ab595a0..d57286f 100644
--- a/src/main/java/net/canarymod/plugin/PluginDescriptor.java
+++ b/src/main/java/net/canarymod/plugin/PluginDescriptor.java
@@ -75,6 +75,9 @@ public class PluginDescriptor {
                     if (zip.getEntry("plugin.yml") != null) {
                         throw new InvalidPluginException("Bukkit Plugins are not natively supported. Please remove '" + pluginFile.getName() + "' from your plugins directory.");
                     }
+                    if (zip.getEntry("mcmod.info") != null) {
+                        throw new InvalidPluginException("Forge Mods or Sponge Plugins are not natively supported. Please remove '" + pluginFile.getName() + "' from your plugins directory.");
+                    }
                     throw new InvalidPluginException("I don't know where to find a Canary.inf in " + path);
                 }
             }
--
2.10.0
