From cabaf492325f57f9eb311e8123725d247c9481ef Mon Sep 17 00:00:00 2001
From: Jamie Mansfield <dev@jamierocks.uk>
Date: Fri, 18 Aug 2017 16:30:50 +0100
Subject: [PATCH] Add performance mixin configuration


diff --git a/src/main/java/net/canarymod/config/ServerConfiguration.java b/src/main/java/net/canarymod/config/ServerConfiguration.java
index 3dc7ea95..6f279ffc 100644
--- a/src/main/java/net/canarymod/config/ServerConfiguration.java
+++ b/src/main/java/net/canarymod/config/ServerConfiguration.java
@@ -185,6 +185,11 @@ public class ServerConfiguration implements ConfigurationContainer {
         cfg.setInt("world-auto-save-interval", 900);
         cfg.setComments("world-auto-save-interval", "The number of ticks between each auto-save");
 
+        // Neptune - start
+        cfg.setBoolean("performance", true);
+        cfg.setComments("performance", "If you want to enable the PERFORMANCE Mixin plugin.");
+        // Neptune - end
+
         cfg.save();
     }
 
@@ -712,4 +717,10 @@ public class ServerConfiguration implements ConfigurationContainer {
     public int getWorldAutoSaveInterval() {
         return this.cfg.getInt("world-auto-save-interval", 900);
     }
+
+    // Neptune - start
+    public boolean getPerformanceEnabled() {
+        return this.cfg.getBoolean("performance", true);
+    }
+    // Neptune - end
 }
-- 
2.14.1

